text/0063-sdk-crash-monitoring.md

```diff
@@ -0,0 +1,87 @@
+
+For some SDKs, like mobile SDKs, we primarily rely on users to report  SDK crashes because we don't operate them in production. If users don't report them, we are unaware. Instead, we should detect crashes caused by our SDKs when they happen so we can proactively fix them.
+
+### SDK Crash Health
+
+In Looker, we could calculate the ratio of crashes reported by each SDK versus the crashes caused by itself to have a metric for how many crashes each SDK causes.
+
+## Background
+
+The Google Play SDK Console provides insights into crashes for SDK maintainers. We regularly use it for the Android/Java SDK. While it would be great also to build something similar for SDK maintainers within Sentry, it's a bit complicated cause of PII and such. Narrowing down the scope to only Sentry SDKs makes the problem easier to solve.
+
+## Options Considered
+
+For every solution, the server or the SDK has to strip all irrelevant data for us to have enough information to solve an SDK crash to reduce privacy implications and risk. They should strip stacktrace frames from our customers, remove most of the context, etc.
+
+- [Option 1: Detect during event processing](#option-1)
+- [Option 2: Detect in SDKs](#option-2)
+- [Option 3: Client Reports](#option-3)
+
+### Option 1: Detect during event processing <a name="option-1"></a>
+
+During event processing, after processing the stacktrace, the server detects if a crash stems from any of our SDKs by looking at the top frames of the stacktrace. If the server detects that it does, it duplicates the event and stores it in a special-cased sentry org where each SDK gets its project.
+
+A good candidate to add this functionality is the `event_manager`. Similarly, where we call [`_detect_performance_problems`](https://github.com/getsentry/sentry/blob/4525f70a1fb521445bbb4c9250b2e15e05b059c3/src/sentry/event_manager.py#L2461), we could add an extra function called `detect_sdk_crashes`.
+
+### Pros <a name="option-1-pros"></a>
+
+1. No need for per-SDK rollout.
+2. No extra events or data sent to Sentry.
+
+### Cons <a name="option-1-cons"></a>
+
+1. Please add your cons.
+
+### Option 2: Detect in SDKs <a name="option-2"></a>
+
+When the SDK sends a crash event to Sentry, it checks the stacktrace and checks if the crash stems from the SDK itself by looking at the top frames of the stacktrace. If it does, the SDK also sends the event to a special-cased sentry org, where each SDK gets its project.
+
+### Cons <a name="option-2-cons"></a>
+
+1. Opposite of [Pro 1 of option 1](#option-1-pros).
+2. Extra events sent from SDKs to Sentry.
+3. Changing the DSN needs an SDK update.
+4. The SDK might end up in an endless loop, if there is a bug in this functionality.
+
+### Option 3: Client Reports <a name="option-3"></a>
```

Comment:
This option feels like the best one to me. 

The client reports dev docs state "Client reports are a protocol feature that let clients send status reports about themselves to Sentry". And a crash is clearly a status report of a Sentry Client.

Reply:
Does Option 3 have any advantages over Option 2? The nice thing about Option 2 is that it would not require any backend changes.

Reply:
Not all SDKs implemented Client reports just yet, and in PHP we will likely never do it.

Reply:
> Does Option 3 have any advantages over Option 2

Yes, the SDKs don't need to send extra events to Sentry. Client reports are attached to already outgoing events.